cincinnati/src/lib.rs

```diff
@@ -187,26 +188,31 @@ impl Graph {
 
     /// Add a transition (edge) from `source` to `target`.
     ///
-    /// Fails with the `WoulcCycle` error if the new edge would lead to a cycle.
-    pub fn add_edge(&mut self, from: &ReleaseId, to: &ReleaseId) -> Result<(), Error> {
+    /// Fails with the `WouldCycle` error if the new edge would lead to a cycle.
+    pub fn add_edge(&mut self, from: &ReleaseId, to: &ReleaseId) -> Result<EdgeIndex, Error> {
+        let from_release: String = self.find_by_releaseid(from)?.version().to_string();
+        let to_release: String = self.find_by_releaseid(to)?.version().to_string();
         if self.dag.find_edge(from.0, to.0).is_some() {
             return Err(Error::from(errors::EdgeAlreadyExists {
-                from: self.find_by_releaseid(from)?.version().to_string(),
-                to: self.find_by_releaseid(to)?.version().to_string(),
+                from: from_release,
+                to: to_release,
             }));
         }
 
         self.dag
             .add_edge(from.0, to.0, Empty {})
-            .map(|_| ())
             .map_err(Into::into)
     }
 
     /// Add edges for all given key/value pairs of releases.
-    pub fn add_edges(&mut self, indices: HashMap<ReleaseId, ReleaseId>) -> Result<(), Error> {
+    pub fn add_edges(
+        &mut self,
+        indices: HashMap<ReleaseId, ReleaseId>,
+    ) -> Result<Vec<EdgeIndex>, Error> {
         indices
             .iter()
-            .try_fold((), |_, (from, to)| self.add_edge(&from, &to))
+            .map(|(from, to)| self.add_edge(&from, &to))
```

Comment:
Given the above comment this could be 

```suggestion
            .map(|(from, to)| (self.add_edge(&from, &to), from, to))
```

I think that the ordering matters here due to the evaluation precedence. Borrows of `to` and `from` are passed to `self.add_edge`, and then their ownership is passed along in the tuple. Besides this thought I didn't test this suggestion, but you'll have to reconcile with the compiler anyway :laughing:

Reply:
Its not gonna work:
```
a value of type `std::result::Result<std::vec::Vec<(petgraph::graph_impl::EdgeIndex, ReleaseId, ReleaseId)>, failure::Error>` cannot be built from an iterator over elements of type `(std::result::Result<petgraph::graph_impl::EdgeIndex, failure::Error>, ReleaseId, ReleaseId)`

value of type `std::result::Result<std::vec::Vec<(petgraph::graph_impl::EdgeIndex, ReleaseId, ReleaseId)>, failure::Error>` cannot be built from `std::iter::Iterator<Item=(std::result::Result<petgraph::graph_impl::EdgeIndex, failure::Error>, ReleaseId, ReleaseId)>
```

ReleaseId's need to be injected in Result<EdgeIndex> somehow

Reply:
Well, it's going to work with the map statement and the changed result

Reply:
To elaborate: `collect()` without a given type will try to built a collection which it finds in whatever it's assigned to, in this case that's the return value for the function. So in this case the `map` statement has to return items which match the return type of the function.